py/desisim/survey_release.py

```diff
@@ -112,128 +98,171 @@ def apply_redshift_dist(self,zmin=0,zmax=10,distribution='SV'):
-        mock_area=SurveyRelease.get_catalog_area(self.mockcatalog,nside=16)
-        dndz_mock=np.histogram(self.mockcatalog['Z_QSO_RSD'],bins=zbins,weights=np.repeat(1/mock_area,len(self.mockcatalog)))[0]
-        fractions=np.divide(dndz,dndz_mock,where=dndz_mock>0,out=np.zeros_like(dndz_mock))
+        
+        selection = self.sample_redshift(dndz,zbins=zbins)
+        selected_ids = list(self.mockcatalog['MOCKID'][selection])
+        self.mockcatalog['IS_QSO_TARGET'] = np.isin(self.mockcatalog['MOCKID'],selected_ids)
+        log.info(f"Selecting {sum(selection)} mock QSO targets following the distribution: {distribution}")
+        if self.include_nonqso_targets:
+            avail_ids = self.mockcatalog['MOCKID'][~selection]
+            dndz_data_nonqso = self.get_nonqso_dndz(expected_dist=dist,zbins=zbins) 
+            selected_nonqso = self.sample_redshift(dndz_data_nonqso,zbins=zbins,mask=~selection)
+            selected_nonqso_ids = avail_ids[selected_nonqso]
+            log.info(f"Selecting {sum(selected_nonqso)} mock non QSO targets following the distribution of data.")
+            self.are_nonqso_targets = np.isin(self.mockcatalog['MOCKID'],selected_nonqso_ids)
+            selection |= self.are_nonqso_targets  
+        log.info(f"Selected {sum(selection)} mock targets out of {len(self.mockcatalog)}")
+        self.mockcatalog=self.mockcatalog[selection]
+                                 
+    def get_nonqso_dndz(self,expected_dist, zbins=np.linspace(0,10,100+1)):
+        """Get the redshift distribution of non QSO targets in the data catalog.
+
+        Args:
+            expected_dist (Table): Table with expected redshift distribution.
+            zbins (array): redshift bins.
+
+        Returns:    
+            dndz_data_nonqso (array): redshift distribution of non QSO targets in the data catalog.
+        """
+        log.info(f"Sampling mock targets from data non QSO targets. Assuming data QSO targets already follow expected distribution.")
+        targetmask = get_qsotarget_mask(self.data)
+        nonqso_targets = self.data[~targetmask]
+        dndz_data_nonqso = np.histogram(nonqso_targets['Z'],bins=zbins)[0]
+        # Figure out the area of the catalog by checking number of Lya. 
+        # This assumes that data lya qso targets already follow the expected distribution.
+        qso_targets = self.data[targetmask]
+        lyabins = np.linspace(2.1,3.8,18)
+        dndz_data_qso = np.histogram(qso_targets['Z'],bins=lyabins)[0]
+        dz_dist=expected_dist['z'][1]-expected_dist['z'][0]
+        factor=0.1/dz_dist # Account for redshift bin width difference.
+        zcenters=0.5*(lyabins[1:]+lyabins[:-1])
+        expected_dndz=factor*np.interp(zcenters,expected_dist['z'],expected_dist['dndz_23'],left=0,right=0)
+        area = np.cumsum(dndz_data_qso)[-1]/np.cumsum(expected_dndz)[-1] 
```

Comment:
If you need the sum of `dndz_data_qso` and `expected_dndz`, you should use `np.sum` rather than `cumsum[-1]`. `np.sum` is smarter than `cumsum[-1]` and reduces the roundoff error (https://numpy.org/doc/stable/reference/generated/numpy.cumsum.html).